### PR TITLE
Add option for ExpansionTile to maintain the state of its children when collapsed

### PR DESCRIPTION
**Pull request by eospi**
## Description
Currently, `ExpansionTile` rebuilds its children upon every expansion and removes the children when collapsed, resetting any state contained in the children. When building a Flutter app a couple months ago, I encountered this issue when placing a `TextFormField` inside an ExpansionTile, and any edited text would be deleted when the user collapsed the ExpansionTile. #37520 further describes the specific issue and includes sample code to reproduce it. I made this change to the `ExpansionTile` widget to make my app work correctly.
### Changes overview:
- Added a `maintainsState` property to `ExpansionTile` of type `bool`
- Set `maintainsState` default value to `false` for the existing behavior
- Add `!maintainsState` as a condition for setting the child of the `AnimatedBuilder` to `null`

Now, an `ExpansionTile` like this maintains the state of its children while collapsed:
```dart
ExpansionTile(
  maintainsState: true,
  key: PageStorageKey("test"),
  title: Text("Click to expand"),
  children: [
    TextFormField(
      key: PageStorageKey("test2"),
      validator: (value) {
        if (value.isEmpty) {
          return 'Please enter some text';
        }
        return null;
      },
    ),
  ]
)
```

## Related Issues

#37520

## Tests

I added the following tests:
- 'ExpansionTile maintainsState' - creates two collapsed ExpansionTile widgets with different `maintainsState` values and tests if a child widget is present in each.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 1]
```diff
@@ -40,6 +40,7 @@ class ExpansionTile extends StatefulWidget {
     this.children = const <Widget>[],
     this.trailing,
     this.initiallyExpanded = false,
+    this.maintainsState = false,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/expansion_tile.dart]: nit: in other places of the framework we call a similar setting "maintainState" without the "s". Can you rename this to be consistent?

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 2]
```diff
@@ -74,6 +75,13 @@ class ExpansionTile extends StatefulWidget {
   /// Specifies if the list tile is initially expanded (true) or collapsed (false, the default).
   final bool initiallyExpanded;
 
+  /// Specifies if the state of the children is maintained when the tile expands and collapses
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/expansion_tile.dart]: nit: sentences should end with a `.`

Comment (goderbauer) [inline packages/flutter/lib/src/material/expansion_tile.dart]:
```suggestion
  /// Specifies whether the state of the children is maintained when the tile is collapsed.
```

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 3]
```diff
@@ -197,10 +205,11 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
   @override
   Widget build(BuildContext context) {
     final bool closed = !_isExpanded && _controller.isDismissed;
+    final bool shouldRemoveChildren = closed && !widget.maintainsState;
     return AnimatedBuilder(
       animation: _controller.view,
       builder: _buildChildren,
-      child: closed ? null : Column(children: widget.children),
+      child: shouldRemoveChildren ? null : Column(children: widget.children),
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/expansion_tile.dart]: Just keeping them in the tree like this doesn't seem like a good idea. Animations will keep running and drain the user's battery even though nothing is visible. We should turn animations off and maybe even put the non-visible children offstage to ensure that you can't interact with them. For an example, look at how the "Visibility" widget implements the maintainState flag.

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 4]
```diff
@@ -215,4 +215,43 @@ void main() {
...
+                  ],
+                ),
+              ],
+            ),
+          ),
+        ),
+    ));
+
+    // This text should be there while ExpansionTile collapsed
+    expect(find.text('Maintaining State'), findsOneWidget);
```
Comment (goderbauer) [inline packages/flutter/test/material/expansion_tile_test.dart]: The tester shouldn't find this if it is not visible unless `skipOffstage: true` is specified.

**Review by goderbauer — CHANGES_REQUESTED** [review]

**Comment by goderbauer** [comment]
Can you also rebase these changes to the latest master? That should make (most of) the tests happy, that are currently failing.

**Comment by eospi** [comment]
Thanks for looking over the PR! I will begin working on the changes as soon as possible.

File: packages/flutter/lib/src/material/expansion_tile.dart (hunk 3 above)
Comment (eospi) [reply packages/flutter/lib/src/material/expansion_tile.dart]: Thanks for pointing to the Visibility widget! Should I wrap the children in a Visibility widget or instead wrap it in both Offstage and TickerMode?

**Comment by goderbauer** [comment]
Hello @eospi,

it looks like maybe a merge or rebase has gone wrong in this PR. It now includes a lot of seemingly unrelated changes. Can you fix this please before I take another look at the code? Also, the analyzer is reporting issues with trailing white space: https://github.com/flutter/flutter/pull/41886/checks?check_run_id=456750789

Thanks!

**Comment by eospi** [comment]
Is it alright if I open a new PR to fix the extra commit issue?

**Comment by goderbauer** [comment]
Go for it! Please close this one when you do.

**Comment by eospi** [comment]
I've just opened #51709 as the replacement PR. Thanks!

**Pull request closed without merging**